// ionic-audio/dist/ionic-audio-providers.js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Injectable } from '@angular/core';
import { WebAudioTrack } from './ionic-audio-web-track';
import { CordovaAudioTrack } from './ionic-audio-cordova-track';
/**
 * Creates an audio provider based on the environment.
 * If running from within a browser, then defaults to HTML5 Audio. If running on a device, it will check for Cordova and Media plugins and use
 * a native audio player, otherwise falls back to HTML5 audio.
 *
 * @method factory
 * @static
 * @return {IAudioProvider} An IAudioProvider instance
 */
export function audioProviderFactory() {
    return window.hasOwnProperty('cordova') && window.hasOwnProperty('Media') ? new CordovaMediaProvider() : new WebAudioProvider();
}
/**
 * Base class for audio providers
 *
 * @export
 * @abstract
 * @class AudioProvider
 * @implements {IAudioProvider}
 */
export var AudioProvider = (function () {
    function AudioProvider() {
    }
    /**
     * Creates an IAudioTrack instance from a JSON object.
     * Not implemented in base class.
     *
     * @method create
     * @param {ITrackConstraint} track A JSON object containing at least a src property
     * @return null
     */
    AudioProvider.prototype.create = function (track) {
        console.error('Not implemented in base class');
        return null;
    };
    /**
     * Adds an existing IAudioTrack instance to the array of managed tracks.
     *
     * @method add
     * @param {IAudioTrack} audioTrack An instance of IAudioTrack
     */
    AudioProvider.prototype.add = function (audioTrack) {
        AudioProvider.tracks.push(audioTrack);
    };
    ;
    /**
     * Plays a given track.
     *
     * @method play
     * @param {number} index The track id
     */
    AudioProvider.prototype.play = function (index) {
        if (index === undefined || index > AudioProvider.tracks.length - 1)
            return;
        this._current = index;
        AudioProvider.tracks[index].play();
    };
    ;
    /**
     * Pauses a given track.
     *
     * @method pause
     * @param {number} [index] The track id, or if undefined it will pause whichever track currently playing
     */
    AudioProvider.prototype.pause = function (index) {
        if (this._current === undefined || index > AudioProvider.tracks.length - 1)
            return;
        index = index || this._current;
        AudioProvider.tracks[index].pause();
    };
    ;
    /**
     * Stops a given track.
     *
     * @method stop
     * @param {number} [index] The track id, or if undefined it will stop whichever track currently playing
     */
    AudioProvider.prototype.stop = function (index) {
        if (this._current === undefined || index > AudioProvider.tracks.length - 1)
            return;
        index = index || this._current;
        AudioProvider.tracks[index].stop();
        this._current = undefined;
    };
    ;
    Object.defineProperty(AudioProvider.prototype, "tracks", {
        /**
         * Gets an array of tracks managed by this provider
         *
         * @property tracks
         * @readonly
         * @type {IAudioTrack[]}
         */
        get: function () {
            return AudioProvider.tracks;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AudioProvider.prototype, "current", {
        /**
         * Gets current track id
         *
         * @property current
         * @type {number}
         */
        get: function () {
            return this._current;
        },
        /**
         * Sets current track id
         *
         * @property current
         */
        set: function (v) {
            this._current = v;
        },
        enumerable: true,
        configurable: true
    });
    AudioProvider.tracks = [];
    AudioProvider = __decorate([
        Injectable(), 
        __metadata('design:paramtypes', [])
    ], AudioProvider);
    return AudioProvider;
}());
/**
 * Creates an HTML5 audio provider
 *
 * @export
 * @class WebAudioProvider
 * @constructor
 * @extends {AudioProvider}
 */
export var WebAudioProvider = (function (_super) {
    __extends(WebAudioProvider, _super);
    function WebAudioProvider() {
        _super.call(this);
        console.log('Using Web Audio provider');
    }
    WebAudioProvider.prototype.create = function (track) {
        var audioTrack = new WebAudioTrack(track.src, track.preload);
        Object.assign(audioTrack, track);
        var trackId = WebAudioProvider.tracks.push(audioTrack);
        audioTrack.id = trackId - 1;
        return audioTrack;
    };
    WebAudioProvider = __decorate([
        Injectable(), 
        __metadata('design:paramtypes', [])
    ], WebAudioProvider);
    return WebAudioProvider;
}(AudioProvider));
/**
 * Creates a Cordova audio provider
 *
 * @export
 * @class CordovaMediaProvider
 * @constructor
 * @extends {AudioProvider}
 */
export var CordovaMediaProvider = (function (_super) {
    __extends(CordovaMediaProvider, _super);
    function CordovaMediaProvider() {
        _super.call(this);
        console.log('Using Cordova Media provider');
    }
    CordovaMediaProvider.prototype.create = function (track) {
        var audioTrack = new CordovaAudioTrack(track.src);
        Object.assign(audioTrack, track);
        var trackId = CordovaMediaProvider.tracks.push(audioTrack);
        audioTrack.id = trackId - 1;
        return audioTrack;
    };
    CordovaMediaProvider = __decorate([
        Injectable(), 
        __metadata('design:paramtypes', [])
    ], CordovaMediaProvider);
    return CordovaMediaProvider;
}(AudioProvider));
